app/routes.py

```diff
@@ -1,2 +1,348 @@
+
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def task_put(task_id):
+    task = Task.query.get_or_404(task_id)
+    form_data = request.get_json()
+
+    task.title = form_data["title"]
+    task.description = form_data["description"]
+
+    db.session.commit()
+    response_body = {"task": task.to_dict()}
+    return jsonify(response_body), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def task_delete(task_id):
+    task = Task.query.get_or_404(task_id)
+    db.session.delete(task)
+    db.session.commit()
+
+    return jsonify({
+        "details": f"Task {task.task_id} \"{task.title}\" successfully deleted"
+    }), 200
+
+
+# *****  Wave 3 (PATCH for complete and incomplete task)  *********
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def task_handle_incomplete_task(task_id):
+    task = Task.query.get_or_404(task_id)
+    task.completed_at = None
+    db.session.commit()
+
+    response_body = {"task": task.to_dict()}
+    return jsonify(response_body), 200
+
+# *****  Wave 3 (PATCH) & 4 (Slack's Bot)  *********
+
+
+def task_slack_bot(title):
+    query_path = {"channel": 'task-notifications', "text": title}
+    header = {"Authorization": f"{os.environ.get('BOT')}"}  # metadata
+    response = requests.post('https://slack.com/api/chat.postMessage',
+                            params=query_path, headers=header)  # (url, param, headers)
+    print("Check to see if this works?")
+    return response.json()  # return the response in json
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def task_handle_complete_task(task_id):
+    task = Task.query.get_or_404(task_id)
+    task.completed_at = datetime.utcnow()
+
+    db.session.commit()
+    task_slack_bot(task.title)
+
+    response_body = {"task": task.to_dict()}
+    return jsonify(response_body), 200
+
+
+# *****  Wave 5  *********
+@goals_bp.route("", methods=["POST"])
+def goal_create():
+    request_body = request.get_json()
+    if "title" not in request_body:
+        return jsonify({"details": "Invalid data"}), 400
+
+    # collecting the value(s)
+    new_goal = Goal(title=request_body["title"])
+
+    # add and update database
+    db.session.add(new_goal)
+    db.session.commit()
+
+    # return the response to our client
+    response_body = {"goal": new_goal.dict()}
+    return jsonify(response_body), 201
+
+
+@goals_bp.route("", methods=["GET"])
+def handle_goals():
+    goals = Goal.query.all()
+    response_body = [goal.dict() for goal in goals]
+    return jsonify(response_body), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+def goal_get(goal_id):
+    goal = Goal.query.get_or_404(goal_id)  # providing the task_id
+    response_body = {"goal": goal.dict()}
+    return jsonify(response_body), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+def goal_put(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+    form_data = request.get_json()
+
+    goal.title = form_data["title"]
```

Comment:
We should also confirm we have valid data within the request body. Similar checks from your `POST` requests would be useful for `PUT` requests too.